File: ios/SwipePhysics.swift
import Foundation
#if canImport(CoreGraphics)
import CoreGraphics
#endif

struct SwipePhysics {

    static func applyRubberBand(
        translation: CGFloat,
        actionsWidth: CGFloat,
        isLeading: Bool
    ) -> CGFloat {
        let divisor = SwipeAnimationConfig.rubberBandDivisor

        if isLeading {
            if translation > actionsWidth {
                return actionsWidth + ((translation - actionsWidth) / divisor)
            } else if translation < 0 {
                return translation / divisor
            }
        } else {
            if translation < -actionsWidth {
                return -actionsWidth - ((abs(translation) - actionsWidth) / divisor)
            } else if translation > 0 {
                return translation / divisor
            }
        }
        return translation
    }

    static func shouldSnapToOpen(
        projectedTranslation: CGFloat,
        actionsWidth: CGFloat,
        threshold: CGFloat,
        isLeading: Bool,
        isCurrentlyOpen: Bool
    ) -> Bool {
        let thresholdDistance = actionsWidth * threshold

        if isLeading {
            return isCurrentlyOpen
                ? projectedTranslation > -thresholdDistance
                : projectedTranslation > thresholdDistance
        } else {
            return isCurrentlyOpen
                ? projectedTranslation < thresholdDistance
                : projectedTranslation < -thresholdDistance
        }
    }

    static func projectFinalPosition(
        translation: CGFloat,
        velocity: CGFloat,
        friction: CGFloat
    ) -> CGFloat {
        let factor = SwipeAnimationConfig.velocityProjectionFactor
        return (translation * friction) + (velocity * friction * factor)
    }

    static func calculateSpringVelocity(from gestureVelocity: CGFloat) -> CGFloat {
        guard abs(gestureVelocity) > 0 else { return 0 }
        let normalizer = SwipeAnimationConfig.velocityNormalizer
        let maxVelocity = SwipeAnimationConfig.maxNormalizedVelocity
        return min(abs(gestureVelocity) / normalizer, maxVelocity)
    }
}

File: android/src/main/java/expo/modules/swipeable/SwipePhysics.kt
package expo.modules.swipeable

import kotlin.math.abs
import kotlin.math.min
import kotlin.math.sign

object SwipePhysics {

    fun applyRubberBand(
        translation: Float,
        actionsWidth: Float,
        isLeading: Boolean
    ): Float {
        var result = translation
        if (isLeading) {
            when {
                result > actionsWidth -> {
                    result = actionsWidth + ((result - actionsWidth) / SwipeConstants.RUBBER_BAND_DIVISOR)
                }
                result < 0f -> {
                    result /= SwipeConstants.RUBBER_BAND_DIVISOR
                }
            }
        } else {
            when {
                result < -actionsWidth -> {
                    result = -actionsWidth - ((abs(result) - actionsWidth) / SwipeConstants.RUBBER_BAND_DIVISOR)
                }
                result > 0f -> {
                    result /= SwipeConstants.RUBBER_BAND_DIVISOR
                }
            }
        }
        return result
    }

    fun shouldSnapToOpen(
        projectedPosition: Float,
        thresholdDistance: Float,
        isOpen: Boolean,
        isLeading: Boolean
    ): Boolean {
        return if (isLeading) {
            if (isOpen) projectedPosition > -thresholdDistance else projectedPosition > thresholdDistance
        } else {
            if (isOpen) projectedPosition < thresholdDistance else projectedPosition < -thresholdDistance
        }
    }

    fun projectFinalPosition(
        translation: Float,
        velocity: Float,
        friction: Float
    ): Float {
        return (translation * friction) + (velocity * friction * SwipeConstants.VELOCITY_PROJECTION_FACTOR)
    }

    fun calculateSpringVelocity(gestureVelocity: Float): Float {
        // Preserve the sign of the velocity - critical for correct spring animation direction
        val magnitude = min(
            abs(gestureVelocity) / SwipeConstants.VELOCITY_NORMALIZER,
            SwipeConstants.MAX_NORMALIZED_VELOCITY
        ) * SwipeConstants.VELOCITY_SCALE
        return sign(gestureVelocity) * magnitude
    }
}
